### DISCUSSION TITLE
[@types/node] Drops support for TypeScript < 3.7 in 12.20.21 ("assert.d.ts:2:68 - error TS1144: '{' or ';' expected.")

### DISCUSSION THREAD

**Discussion by Chealer** [DefinitelyTyped/DefinitelyTyped#55429] [Issues with a @types package]
Our TypeScript 3.5 project stopped compiling yesterday, due to an upgrade to @types/node 12.20.21, with a bunch of errors:

> ERROR in ../node_modules/@types/node/assert.d.ts:2:68 - error TS1144: '{' or ';' expected.
> 2     function assert(value: any, message?: string | Error): asserts value;

![image](https://user-images.githubusercontent.com/486141/131140254-b8827120-05b7-4cc9-aeb3-55887f1425eb.png)

This would be no surprise as TypeScript 3.5 is fairly old and DefinitelyTyped does not support old versions indefinitely, except after analysis it turns out our package.json properly specific the dependency:
`    "@types/node": "^12.7.1",`

In fact, @types/node had already dropped support for TypeScript 3.6, but for some reason, this week this change was backported to branch 12, and while 12.20.20 was fine, 12.20.21 drops support, breaking semantic versioning.

[a comment by an automated account is not shown]

**Reply by peterblazejewicz** [reply]
@SimonSchick this could be related to change merged yesterday, mind providing input? appreciated, thx!

**Reply by SimonSchick** [reply]
A couple notes:
- DT does not really follow semver as `major version of types` = `major version of package`.
- DT *does* have it's own TS support windows which did expire prior this change

I agree that deleting the v12 ts3.6 folder may have been a little pre-mature but did not violate any DT guidelines, 

So the options here are:
- Ask people to pin version of node v12 types or ask them to roll a new TS version, going from ts3.6 to 3.7 isn't a big change, but some orgs may be resistant/hesitant.
- Revert v12 types to previous state, I'm unsure if dtslint will still work w/ 3.6.

I will draft a PR to revert it for v12 and see what happens.

**Reply by Chealer** [reply]
Thank you @SimonSchick 

Note that the folder, although called 3.6, was visibly used for TypeScript 3.5 too. 
I am new to TypeScript, but unless we can have several TypeScript versions, my understanding is that migrating from 3.5 to 3.7 would force us to migrate to Angular 9 (which I consider as a sizable project).

**Reply by SimonSchick** [reply]
In that case you might have to freeze your node typings, support for TS3.5 has ended a while ago.

**Reply by sandersn** [reply]
> * Ask people to pin version of node v12 types.

This is the only option that will work; DT tries to layer *two* versions into a single version, so it can't truly support semantic versioning. Specifically, it uses the patch version for every single change that's published, whether or not it's breaking.

@Chealer if you're outside the 2-year DT support window, you need to specify exact versions of all `@types` packages. (You should probably do this with everything -- a project on two-year-old dependencies should not be upgrading to the latest version of anything.)

@SimonSchick even if dtslint allowed ts3.6 folders, the publisher would also have to publish them again.

**Comment by AlonsoM45** [comment]
Hi, my team is also experiencing this issue since yesterday.

**Comment by aespinoza96** [comment]
All the builds that are using this version are experiencing the same problem, I think this is very critical.
